File: Algorithms/A Star Search.cpp
class graph {
    unordered_map<string, unordered_map<string, int>> edge;
public:
    void addPoint(string p) {
        if (edge.find(p) == edge.end())
            this->edge[p] = {};
    }
    void addEdge(string p1, string p2, int weight) {
        if (edge.find(p1) != edge.end() && edge.find(p2) != edge.end()) {
            this->edge[p1].emplace(p2, weight);
            this->edge[p2].emplace(p1, weight);
        }
    }
    unordered_map<string, int> getEdge(string p) {
        return edge[p];
    }
    int getWeight(string from,string to) {
        if (edge[from].find(to) != edge[from].end()) {
            return edge[from][to];
        }
        return -1;
    }
    bool hasPoint(string p) {
        return edge.find(p) != edge.end();
    }

    unordered_map<string, int> operator[](string p) {
        return this->edge[p];
    }
};

bool cmp(pair<string, int> left, pair<string, int> right) {
    return left.second < right.second;
}
pair<double, vector<string>> aStar(graph g, unordered_map<string, int> h, string begin, string end) {
    // f(n) = g(n) + h(n)
    if (!(g.hasPoint(begin) && g.hasPoint(end))) return { -1,{} };
    if (begin == end) return { 0,{begin, end} };

    int len = 0;
    vector<string> road;
    road.emplace_back(begin);

    string cur = begin;
    while (cur != end) {
        vector<pair<string, int>> f;
        unordered_map<string, int> temp = g[cur];
        auto ib = temp.begin();
        auto ie = temp.end();
        while (ib != ie) {
            // f.emplace_back(ib->first, ib->second + h[ib->first]);
            f.emplace_back(ib->first, len + ib->second + h[ib->first]);
            ib++;
        }
        sort(f.begin(), f.end(), cmp);
        len += g.getWeight(cur, f[0].first);
        road.emplace_back(f[0].first);
        cur = f[0].first;
    }
    return { len,road };
}

File: Algorithms/A Star Search.py
class graph:
    edge: Dict[str, Dict[str, int]]

    def __init__(self):
        self.edge = {}
        self.edge.clear()

    def addpoint(self, point: str):
        if point not in self.edge:
            self.edge.setdefault(point, {})

    def addedge(self, p1: str, p2: str, weight: int):
        if p1 in self.edge and p2 in self.edge:
            self.edge[p1][p2] = weight
            self.edge[p2][p1] = weight

    def getedge(self, point: str) -> Dict[str, int]:
        return self.edge[point]

    def getweight(self, p_from: str, p_to: str):
        if p_to in self.edge[p_from]:
            return self.edge[p_from][p_to]
        return -1


def astar(g: graph, h: Dict[str, int], begin: str, end: str) -> tuple:
    if begin == end:
        return 0, (begin, end)

    length: int = 0
    road: tuple = (begin,)

    cur: str = begin
    while cur != end:
        f: List[tuple] = []
        temp: Dict[str, int] = g.getedge(cur)
        for k, v in temp.items():
            f += [(k, length + v + h[k])]
        newf: List[tuple] = sorted(f, key=lambda a: a[1])
        length += g.getweight(cur, newf[0][0])
        road += (newf[0][0],)
        cur = newf[0][0]
    return length, road
